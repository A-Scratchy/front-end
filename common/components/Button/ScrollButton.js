import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import ReactGA from 'react-ga';
import { Link as ScrollLink, Events as ScrollEvent } from 'react-scroll';
import styles from './Button.css';

ScrollButton.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  fullWidth: PropTypes.bool,
  href: PropTypes.string,
  onClick: PropTypes.func,
  tabIndex: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  theme: PropTypes.oneOf(['primary', 'secondary', 'slate']),
};

ScrollButton.defaultProps = {
  className: '',
  fullWidth: false,
  href: '',
  onClick: () => {},
  tabIndex: 0,
  theme: 'primary',
};

function ScrollButton({
 className, children, fullWidth, href, onClick, tabIndex, theme
}) {
  const buttonClassNames = classNames(
styles.Button, className, {
    [styles.primary]: theme === 'primary',
    [styles.secondary]: theme === 'secondary',
    [styles.slate]: theme === 'slate',
    [styles.fullWidth]: fullWidth,
  }
);

  // TODO: Handle non-string input for analytics event label on both outbound and scroll link
  // Example: SVG as a child
  const eventDetails = {
    category: 'Scroll Button Clicked',
    action: `[${children}] from ${window.location.pathname}`,
  };

  const isProd = process.env.NODE_ENV === 'production';

  // Report scroll link button clicks to Google Analytics
  if (isProd) {
    ScrollEvent.scrollEvent.register('begin', () => {
      ReactGA.event(eventDetails);
    });
  }

  const clickHandler = () => {
    if (!isProd) {
      const analyticsMessage = `Analytics disabled.
        Message: ${eventDetails.category} - ${eventDetails.action}`;

      // eslint-disable-next-line no-console
      console.log(analyticsMessage);
      onClick(analyticsMessage);
      return;
    }

    onClick();
  };

  const onEnterHandler = ev => (ev.key === 'Enter' ? clickHandler : () => {});

  return (
    <ScrollLink
      className={buttonClassNames}
      duration={400}
      onClick={clickHandler}
      onKeyDown={onEnterHandler}
      smooth
      tabIndex={tabIndex}
      to={href}
    >
      {children}
    </ScrollLink>
  );
}

export default ScrollButton;
