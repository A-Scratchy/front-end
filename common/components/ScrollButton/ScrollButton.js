import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import ReactGA from 'react-ga';
import { Link as ScrollLink, Events as ScrollEvent } from 'react-scroll';
import styles from 'common/components/Button/Button.css';

ScrollButton.propTypes = {
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.string]).isRequired,
  className: PropTypes.string,
  fullWidth: PropTypes.bool,
  href: PropTypes.string,
  onClick: PropTypes.func,
  tabIndex: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  theme: PropTypes.oneOf(['primary', 'secondary', 'slate']),
};

ScrollButton.defaultProps = {
  className: '',
  fullWidth: false,
  href: '',
  onClick: undefined,
  tabIndex: 0,
  theme: 'primary',
};

function ScrollButton({ className, children, fullWidth, href, onClick, tabIndex, theme }) {
  const buttonClassNames = classNames(styles.Button, className, {
    [styles.primary]: theme === 'primary',
    [styles.secondary]: theme === 'secondary',
    [styles.slate]: theme === 'slate',
    [styles.fullWidth]: fullWidth,
  });

  const isProd = process.env.NODE_ENV === 'production';

  const clickHandler = () => {
    if (!isProd) {
      // eslint-disable-next-line no-console
      console.log(`Analytics disabled. <ScrollButton> clicked.`);
      return onClick;
    }

    ScrollEvent.scrollEvent.register('begin', () => {
      ReactGA.event({
        category: 'Interactions',
        action: 'Clicked Scroll Button',
        label: `To [${href}]`,
      });
    });

    return onClick;
  };

  const onEnterHandler = ev => {
    if (ev.key === 'Enter') {
      clickHandler();
    }
  };

  return (
    <ScrollLink
      className={buttonClassNames}
      duration={400}
      onClick={clickHandler}
      onKeyDown={onEnterHandler}
      smooth
      tabIndex={tabIndex}
      to={href}
    >
      {/* Render text nodes within a span to apply selector styles */}
      {typeof children === 'string' ? <span>{children}</span> : children}
    </ScrollLink>
  );
}

export default ScrollButton;
