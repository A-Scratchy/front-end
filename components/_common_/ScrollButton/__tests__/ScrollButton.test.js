/* eslint-env jest */
import React from 'react';
import { mount, shallow } from 'enzyme';
import ReactGA from 'react-ga';
import createSnapshotTest from 'test-utils/createSnapshotTest';

import ScrollButton from '../ScrollButton';

describe('ScrollButton', () => {
  it('should render with required props', () => {
    createSnapshotTest(<ScrollButton href="#test">Test</ScrollButton>);
  });

  it('should render with all props assigned', () => {
    createSnapshotTest(
      <ScrollButton
        className="test"
        fullWidth
        href="#test"
        onClick={jest.fn()}
        tabIndex={-1}
        theme="secondary"
      >
        Test
      </ScrollButton>,
    );
  });

  it('should render without a generated span when children is PropTypes.node', () => {
    const testText = 'Testing No Span';

    const ScrollButtonInstance = mount(
      <ScrollButton href="#test">
        <b>{testText}</b>
      </ScrollButton>,
    );

    expect(
      ScrollButtonInstance.containsAnyMatchingElements([
        <span>{testText}</span>,
        <span>
          <b>{testText}</b>
        </span>,
      ]),
    ).toStrictEqual(false);
  });

  it('should render with a generated span when children is PropTypes.string', () => {
    const testText = 'Testing With Span';

    const ScrollButtonInstance = mount(<ScrollButton href="#test">{testText}</ScrollButton>);

    expect(
      ScrollButtonInstance.containsAllMatchingElements([<span>{testText}</span>]),
    ).toStrictEqual(true);
  });

  it('should send log to console when clickHandler is called in non-prod environment', () => {
    /* eslint-disable no-console */
    console.log = jest.fn();

    const ScrollButtonShallowInstance = shallow(<ScrollButton href="#test">Testing</ScrollButton>);

    ScrollButtonShallowInstance.instance().clickHandler();

    expect(console.log.mock.calls).toHaveLength(1);
    /* eslint-enable no-console */
  });

  it('should call ReactGA when in prod environment', () => {
    /* eslint-disable no-console */
    ReactGA.initialize('foo', { testMode: true });

    process.env.NODE_ENV = 'production';

    const ScrollButtonShallowInstance = shallow(<ScrollButton href="#test">Testing</ScrollButton>);

    ScrollButtonShallowInstance.instance().clickHandler();

    expect(ReactGA.testModeAPI.calls).toContainEqual(['create', 'foo', 'auto']);
  });
});
