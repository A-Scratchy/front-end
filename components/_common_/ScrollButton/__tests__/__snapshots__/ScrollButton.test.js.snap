// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollButton should render with all props assigned 1`] = `
<a
  className="Button test secondary fullWidth"
  onClick={[Function]}
  tabIndex={-1}
>
  <span>
    Test
  </span>
</a>
`;

exports[`ScrollButton should render with required props 1`] = `
<a
  className="Button primary"
  onClick={[Function]}
  tabIndex={0}
>
  <span>
    Test
  </span>
</a>
`;
